no-useless-escape should check template literals

`no-useless-escape` does not currently check template literals.

**What version are you using?**
v3.7.1

**What did you do?**
```json
{
  "parserOptions": {
    "ecmaVersion": 6
  },
  "rules": {
    "no-useless-escape": "error"
  }
}
```
```js
var test = `\"${a}\"`;
```
**What happened?**
No errors reported.

**What did you expect to happen?**
```
  1:13  error  Unnecessary escape character: \"  no-useless-escape
  1:19  error  Unnecessary escape character: \"  no-useless-escape

✖ 2 problems (2 errors, 0 warnings)
```

I can put together a PR to fix this.

--- a/docs/rules/no-useless-escape.md
+++ b/docs/rules/no-useless-escape.md
@@ -1,10 +1,11 @@
 # Disallow unnecessary escape usage (no-useless-escape)
 
-Escaping non-special characters in strings and regular expressions doesn't have any effects on results, as in the following example:
+Escaping non-special characters in strings, template literals, and regular expressions doesn't have any effect, as demonstrated in the following example:
 
 ```js
 let foo = "hol\a"; // > foo = "hola"
-let bar = /\:/ // same functionality with /:/
+let bar = `${foo}\!`; // > bar = "hola!"
+let baz = /\:/ // same functionality with /:/
 ```
 
 ## Rule Details
@@ -20,6 +21,9 @@ The following patterns are considered problems:
 '\"';
 "\#";
 "\e";
+`\"`;
+`\"${foo}\"`;
+`\#{foo}`;
 /\!/;
 /\@/;
 
@@ -36,6 +40,9 @@ The following patterns are not considered problems:
 "\u00a9";
 "\371";
 "xs\u2111";
+`\``;
+`\${${foo}\}`;
+`$\{${foo}\}`;
 /\\/g;
 /\t/g;
 /\\w\\$\\*\\^\\./;

--- a/lib/rules/no-useless-escape.js
+++ b/lib/rules/no-useless-escape.js
@@ -76,23 +76,44 @@ module.exports = {
          * @private
          * @param {string[]} escapes - list of valid escapes
          * @param {ASTNode} node - node to validate.
-         * @param {string} elm - string slice to validate.
+         * @param {string} match - string slice to validate.
          * @returns {void}
          */
-        function validate(escapes, node, elm) {
-            const escapeNotFound = escapes.indexOf(elm[0][1]) === -1;
-            const isQuoteEscape = elm[0][1] === node.raw[0];
+        function validate(escapes, node, match) {
+            const isTemplateElement = node.type === "TemplateElement";
+            const escapedChar = match[0][1];
+            let isUnnecessaryEscape = escapes.indexOf(escapedChar) === -1;
+            let isQuoteEscape;
 
-            if (escapeNotFound && !isQuoteEscape) {
+            if (isTemplateElement) {
+                isQuoteEscape = escapedChar === "`";
+
+                if (escapedChar === "$") {
+
+                    // Warn if `\$` is not followed by `{`
+                    isUnnecessaryEscape = match.input[match.index + 2] !== "{";
+                } else if (escapedChar === "{") {
+
+                    /* Warn if `\{` is not preceded by `$`. If preceded by `$`, escaping
+                     * is necessary and the rule should not warn. If preceded by `/$`, the rule
+                     * will warn for the `/$` instead, as it is the first unnecessarily escaped character.
+                     */
+                    isUnnecessaryEscape = match.input[match.index - 1] !== "$";
+                }
+            } else {
+                isQuoteEscape = escapedChar === node.raw[0];
+            }
+
+            if (isUnnecessaryEscape && !isQuoteEscape) {
                 context.report({
                     node,
                     loc: {
                         line: node.loc.start.line,
-                        column: node.loc.start.column + elm.index
+                        column: node.loc.start.column + match.index
                     },
                     message: "Unnecessary escape character: {{character}}.",
                     data: {
-                        character: elm[0]
+                        character: match[0]
                     }
                 });
             }
@@ -105,12 +126,18 @@ module.exports = {
          * @returns {void}
          */
         function check(node) {
-            let nodeEscapes, match;
+            const isTemplateElement = node.type === "TemplateElement";
+            const value = isTemplateElement ? node.value.raw : node.raw;
             const pattern = /\\[^\d]/g;
+            let nodeEscapes,
+                match;
 
-            if (typeof node.value === "string") {
+            if (typeof node.value === "string" || isTemplateElement) {
 
-                // JSXAttribute doesn't have any escape sequence: https://facebook.github.io/jsx/
+                /*
+                 * JSXAttribute doesn't have any escape sequence: https://facebook.github.io/jsx/.
+                 * In addition, backticks are not supported by JSX yet: https://github.com/facebook/jsx/issues/25.
+                 */
                 if (node.parent.type === "JSXAttribute") {
                     return;
                 }
@@ -122,12 +149,14 @@ module.exports = {
                 return;
             }
 
-            while ((match = pattern.exec(node.raw))) {
+            while ((match = pattern.exec(value))) {
                 validate(nodeEscapes, node, match);
             }
         }
+
         return {
-            Literal: check
+            Literal: check,
+            TemplateElement: check
         };
     }
 };

--- a/tests/lib/rules/no-useless-escape.js
+++ b/tests/lib/rules/no-useless-escape.js
@@ -47,7 +47,35 @@ ruleTester.run("no-useless-escape", rule, {
         "var foo = '\\\n';",
         "var foo = '\\\r\n';",
         {code: "<foo attr=\"\\d\"/>", parserOptions: {ecmaFeatures: {jsx: true}}},
-        {code: "<foo attr='\\d'></foo>", parserOptions: {ecmaFeatures: {jsx: true}}}
+        {code: "<foo attr='\\d'></foo>", parserOptions: {ecmaFeatures: {jsx: true}}},
+        {code: "var foo = `\\x123`", parserOptions: {ecmaVersion: 6}},
+        {code: "var foo = `\\u00a9`", parserOptions: {ecmaVersion: 6}},
+        {code: "var foo = `xs\\u2111`", parserOptions: {ecmaVersion: 6}},
+        {code: "var foo = `foo \\\\ bar`;", parserOptions: {ecmaVersion: 6}},
+        {code: "var foo = `\\t`;", parserOptions: {ecmaVersion: 6}},
+        {code: "var foo = `foo \\b bar`;", parserOptions: {ecmaVersion: 6}},
+        {code: "var foo = `\\n`;", parserOptions: {ecmaVersion: 6}},
+        {code: "var foo = `foo \\r bar`;", parserOptions: {ecmaVersion: 6}},
+        {code: "var foo = `\\v`;", parserOptions: {ecmaVersion: 6}},
+        {code: "var foo = `\\f`;", parserOptions: {ecmaVersion: 6}},
+        {code: "var foo = `\\\n`;", parserOptions: {ecmaVersion: 6}},
+        {code: "var foo = `\\\r\n`;", parserOptions: {ecmaVersion: 6}},
+        {code: "var foo = `${foo} \\x123`", parserOptions: {ecmaVersion: 6}},
+        {code: "var foo = `${foo} \\u00a9`", parserOptions: {ecmaVersion: 6}},
+        {code: "var foo = `${foo} xs\\u2111`", parserOptions: {ecmaVersion: 6}},
+        {code: "var foo = `${foo} \\\\ ${bar}`;", parserOptions: {ecmaVersion: 6}},
+        {code: "var foo = `${foo} \\b ${bar}`;", parserOptions: {ecmaVersion: 6}},
+        {code: "var foo = `${foo}\\t`;", parserOptions: {ecmaVersion: 6}},
+        {code: "var foo = `${foo}\\n`;", parserOptions: {ecmaVersion: 6}},
+        {code: "var foo = `${foo}\\r`;", parserOptions: {ecmaVersion: 6}},
+        {code: "var foo = `${foo}\\v`;", parserOptions: {ecmaVersion: 6}},
+        {code: "var foo = `${foo}\\f`;", parserOptions: {ecmaVersion: 6}},
+        {code: "var foo = `${foo}\\\n`;", parserOptions: {ecmaVersion: 6}},
+        {code: "var foo = `${foo}\\\r\n`;", parserOptions: {ecmaVersion: 6}},
+        {code: "var foo = `\\\``", parserOptions: {ecmaVersion: 6}},
+        {code: "var foo = `\\\`${foo}\\\``", parserOptions: {ecmaVersion: 6}},
+        {code: "var foo = `\\${{${foo}`;", parserOptions: {ecmaVersion: 6}},
+        {code: "var foo = `$\\{{${foo}`;", parserOptions: {ecmaVersion: 6}}
     ],
 
     invalid: [
@@ -75,6 +103,59 @@ ruleTester.run("no-useless-escape", rule, {
             code: "<foo attr={\"\\d\"}/>",
             parserOptions: {ecmaFeatures: {jsx: true}},
             errors: [{ line: 1, column: 13, message: "Unnecessary escape character: \\d.", type: "Literal"}]
+        },
+        { code: "var foo = '\\`';", errors: [{ line: 1, column: 12, message: "Unnecessary escape character: \\`.", type: "Literal"}] },
+        { code: "var foo = `\\\"`;", parserOptions: {ecmaVersion: 6}, errors: [{ line: 1, column: 11, message: "Unnecessary escape character: \\\".", type: "TemplateElement"}] },
+        { code: "var foo = `\\'`;", parserOptions: {ecmaVersion: 6}, errors: [{ line: 1, column: 11, message: "Unnecessary escape character: \\'.", type: "TemplateElement"}] },
+        { code: "var foo = `\\#`;", parserOptions: {ecmaVersion: 6}, errors: [{ line: 1, column: 11, message: "Unnecessary escape character: \\#.", type: "TemplateElement"}] },
+        {
+            code: "var foo = '\\\`foo\\\`';",
+            errors: [
+                { line: 1, column: 12, message: "Unnecessary escape character: \\`.", type: "Literal"},
+                { line: 1, column: 17, message: "Unnecessary escape character: \\`.", type: "Literal"},
+            ]
+        },
+        {
+            code: "var foo = `\\\"${foo}\\\"`;",
+            parserOptions: {ecmaVersion: 6},
+            errors: [
+                { line: 1, column: 11, message: "Unnecessary escape character: \\\".", type: "TemplateElement"},
+                { line: 1, column: 19, message: "Unnecessary escape character: \\\".", type: "TemplateElement"},
+            ]
+        },
+        {
+            code: "var foo = `\\\'${foo}\\\'`;",
+            parserOptions: {ecmaVersion: 6},
+            errors: [
+                { line: 1, column: 11, message: "Unnecessary escape character: \\'.", type: "TemplateElement"},
+                { line: 1, column: 19, message: "Unnecessary escape character: \\'.", type: "TemplateElement"}
+            ]
+        },
+        {
+            code: "var foo = `\\#${foo}`;",
+            parserOptions: {ecmaVersion: 6},
+            errors: [{ line: 1, column: 11, message: "Unnecessary escape character: \\#.", type: "TemplateElement"}]
+        },
+        {
+            code: "var foo = `\\$\\{{${foo}`;",
+            parserOptions: {ecmaVersion: 6},
+            errors: [
+                { line: 1, column: 11, message: "Unnecessary escape character: \\$.", type: "TemplateElement"},
+            ]
+        },
+        {
+            code: "var foo = `\\$a${foo}`;",
+            parserOptions: {ecmaVersion: 6},
+            errors: [
+                { line: 1, column: 11, message: "Unnecessary escape character: \\$.", type: "TemplateElement"},
+            ]
+        },
+        {
+            code: "var foo = `a\\{{${foo}`;",
+            parserOptions: {ecmaVersion: 6},
+            errors: [
+                { line: 1, column: 12, message: "Unnecessary escape character: \\{.", type: "TemplateElement"},
+            ]
         }
     ]
 });

